### PR TITLE
[CI] Move to Github actions

### PR DESCRIPTION
**Pull request by summer-alice**
As per your suggestion in #64, I took a look at moving the CI to Github actions.  The result is that the same tasks Travis/AppVeyor ran are replicated here, that is to say it only builds DWT and runs the `tools/test_snippets.d` file.  [Here is the most recent run](https://github.com/summer-alice/dwt/actions/runs/514130969).

`dmd-beta`, `dmd-latest`, and `ldc-latest` are tested on Ubuntu and Windows x86_64, while only `dmd-latest` and `ldc-latest` are tested on Windows x86.  There was no particular reason for the choices so feel free to request changes.

Also, it should be possible to run the unit tests for the `base` package if you want.

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
> dmd-beta, dmd-latest, and ldc-latest are tested on Ubuntu and Windows x86_64, while only dmd-latest and ldc-latest are tested on Windows x86. There was no particular reason for the choices so feel free to request changes.

That's fine. Although, I think it should run for nightly/master as well.

The Travis and AppVeyor were running using a version old version of DMD. That's on purpose. Since this is a library I don't want to put any restrictions, or as few restrictions as possible, on the compiler the user can pick. So, ideally it should test all major versions since 2.078.1. Or we should be a couple of versions, if there are too many to test.

> Also, it should be possible to run the unit tests for the base package if you want.

It should run as many tests as possible.

**Comment by jacob-carlborg** [comment]
I think there should be a scheduled build to run every day. This in combination with testing master/nightly will allow us to catch any regressions/breaking changes in upstream early.

**Comment by summer-alice** [comment]
> > dmd-beta, dmd-latest, and ldc-latest are tested on Ubuntu and Windows x86_64, while only dmd-latest and ldc-latest are tested on Windows x86. There was no particular reason for the choices so feel free to request changes.
> 
> That's fine. Although, I think it should run for nightly/master as well.

Added both `dmd-master` and `ldc-master`.

> The Travis and AppVeyor were running using a version old version of DMD. That's on purpose. Since this is a library I don't want to put any restrictions, or as few restrictions as possible, on the compiler the user can pick. So, ideally it should test all major versions since 2.078.1. Or we should be a couple of versions, if there are too many to test.

I've added `dmd-2.078.3`, `dmd-2.085.1`, and `dmd-2.090.1` with the idea being we test every 5 versions.  From what I can see, there isn't a limit on the number of "tasks" you can run for Github actions, but with this we're already at 18 and no testing for macOS.

> > Also, it should be possible to run the unit tests for the base package if you want.
> 
> It should run as many tests as possible.

I added testing for the base package, however, there is an error! 

```
base/src/java/nonstandard/SafeUtf.d(32,5): Error: foreach: key cannot be of non-integral type UTF8index
base/src/java/nonstandard/SafeUtf.d(69,24): Error: function java.nonstandard.SafeUtf.validateUTF8index(in char[] s, in UTF8index i) is not callable using argument types (string, ulong)
base/src/java/nonstandard/SafeUtf.d(69,24):        cannot pass argument s.length of type ulong to parameter in UTF8index i
/usr/bin/dmd failed with exit code 1.
```

I can take a look at fixing this.

There also seems to be an error running `tools/test_snippets.d` but the logs won't show for me (need to try again later apparently), https://github.com/summer-alice/dwt/runs/1781759665?check_suite_focus=true

File: .github/workflows/ci.yml
Comment (jacob-carlborg) [inline .github/workflows/ci.yml:17]: Please make a vertical array instead. I think this line is getting too long.

**Comment by summer-alice** [comment]
With the latest changes there are some more errors that can be seen:

1. `dmd-2.078.3`: `base/src/java/util/HashSet.d(104,16): Error: no property 'update' for type 'void[0][Object]'`
2. `dmd-2.090.1`:
    ```
    base/src/java/util/HashSet.d(104,23): Error: template `object.update` cannot deduce function from argument types `!()(void[0][Object], Object, void[0] delegate() pure nothrow @nogc @safe, void function(ref void[0] _param_0) pure nothrow @nogc @safe)`, candidates are:
    /opt/hostedtoolcache/dc/dmd-2.090.1/x64/dmd2/linux/bin64/../../src/druntime/import/object.d(2649,6):        `update(K, V, C, U)(ref V[K] aa, K key, scope C create, scope U update)`
      with `K = object.Object,
            V = void[0],
            C = void[0] delegate() pure nothrow @nogc @safe,
            U = void function(ref void[0]) pure nothrow @nogc @safe`
       must satisfy the following constraint:
     `       is(typeof(update(aa[K.init])) : V)`
    ```
3. `dmd-latest`: `core.exception.AssertError@base/src/java/lang/util.d(332): unittest failure`

Number 3 is just a typo, though upon fixing it, the `assert` on [line 348] fails, and I think this would be tough to fix.  
Fixing 2 seems to just need `return (void[0]).init`. 
I'm not sure what you want to do about number 1 though.

[line 348]: https://github.com/d-widget-toolkit/dwt/blob/master/base/src/java/lang/util.d#L348

**Comment by jacob-carlborg** [comment]
> Number 3 is just a typo, though upon fixing it, the assert on line 348 fails, and I think this would be tough to fix.

Why would that be difficult to fix?

> I'm not sure what you want to do about number 1 though.

Yeah, so the `update` function is a fairly recent addition. Either you can go back to what you had before the `update` function. Or, check if the `update` function is available, otherwise fall back to the code you had before.

**Comment by summer-alice** [comment]
> > Number 3 is just a typo, though upon fixing it, the assert on line 348 fails, and I think this would be tough to fix.
> 
> Why would that be difficult to fix?

So I take it back a little.  It likely wouldn't be hard, we'd just have to re-write a bit so the [`fmtFromTangoFmt`](https://github.com/d-widget-toolkit/dwt/blob/master/base/src/java/lang/util.d#L148) function knows when it's dealing with an array and set the default format string be `[%(%` and append `%)]` at the end.  In this particular case, the final string would be `[%(%x, %)]`.  For the `assert` following that: `[%(%-4s, %)]`.

> > I'm not sure what you want to do about number 1 though.
> 
> Yeah, so the `update` function is a fairly recent addition. Either you can go back to what you had before the `update` function. Or, check if the `update` function is available, use that, otherwise fall back to the code you had before.

To check if `update` exists, that would just be `static if (__traits(hasMember, set, "update"))`, correct?

**Comment by jacob-carlborg** [comment]
> To check if update exists, that would just be static if (__traits(hasMember, set, "update")), correct?

No, that won't work. `update` is not a member of an associative array. It's free function declared in the `object` module [1]. It can be called with the member syntax due to D supporting UFCS [2]. You would need to check if the `object` module has an `update` member:

```d
static if (__traits(hasMember, .object, "update"))
```

 Note the dot in front of `object`. This is to make sure the name lookup starts at the module level.

[1] https://dlang.org/phobos/object.html#.update
[2] https://dlang.org/spec/function.html#pseudo-member

**Comment by summer-alice** [comment]
> > To check if update exists, that would just be static if (__traits(hasMember, set, "update")), correct?
> 
> No, that won't work. `update` is not a member of an associative array. It's free function declared in the `object` module [1]. It can be called with the member syntax due to D supporting UFCS [2].
>
> [...]
> 
> [1] https://dlang.org/phobos/object.html#.update
> [2] https://dlang.org/spec/function.html#pseudo-member

Gotcha.

> Note the dot in front of `object`. This is to make sure the name lookup starts at the module level.

I wasn't aware of this, thanks for pointing it out.

**Comment by summer-alice** [comment]
With the merge of #71 some tests pass (yay!) however something that I didn't encounter during local tests is an assertion error in `java.nonstandard.sync.condition` [Line 596].

Since this seems to be taken from an older version of [druntime/sync], and since we're dropping Tango support, is there any reason to not use `core.sync.*` instead?

A quick `grep` shows the following changes would be needed:

```
base/src/java/nonstandard/sync/condition.d:16:module java.nonstandard.sync.condition;
base/src/java/nonstandard/sync/condition.d:24:public import java.nonstandard.sync.exception;
base/src/java/nonstandard/sync/condition.d:25:public import java.nonstandard.sync.mutex;
base/src/java/nonstandard/sync/condition.d:442:    private import java.nonstandard.sync.mutex;
base/src/java/nonstandard/sync/exception.d:15:module java.nonstandard.sync.exception;
base/src/java/nonstandard/sync/mutex.d:16:module java.nonstandard.sync.mutex;
base/src/java/nonstandard/sync/mutex.d:25:public import java.nonstandard.sync.exception;

org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/widgets/RunnableLock.d:22:    import java.nonstandard.sync.condition;
org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/widgets/RunnableLock.d:23:    import java.nonstandard.sync.mutex;
org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/widgets/Synchronizer.d:26:    import java.nonstandard.sync.exception: SyncException;
org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/internal/Lock.d:22:    import java.nonstandard.sync.mutex;
org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/internal/Lock.d:23:    import java.nonstandard.sync.condition;

org.eclipse.swt.win32.win32.x86/src/org/eclipse/swt/widgets/RunnableLock.d:21:    import java.nonstandard.sync.condition;
org.eclipse.swt.win32.win32.x86/src/org/eclipse/swt/widgets/RunnableLock.d:22:    import java.nonstandard.sync.mutex;
org.eclipse.swt.win32.win32.x86/src/org/eclipse/swt/widgets/Synchronizer.d:28:    import java.nonstandard.sync.exception;
org.eclipse.swt.win32.win32.x86/src/org/eclipse/swt/internal/Lock.d:21:    import java.nonstandard.sync.mutex;
org.eclipse.swt.win32.win32.x86/src/org/eclipse/swt/internal/Lock.d:22:    import java.nonstandard.sync.condition;
```

The first set are a non-issue since that's the code that would be replaced.  A look through the other files and there isn't any API breakages, so it'd just be different imports.

[Line 596]: https://github.com/d-widget-toolkit/dwt/blob/49a1f5401c7bda28b878511c9ff11a05b10583c5/base/src/java/nonstandard/sync/condition.d#L596
[druntime/sync]: https://github.com/dlang/druntime/tree/d4487d8931a42547f296773ecb8d576cda8a2f24/src/core/sync

**Comment by jacob-carlborg** [comment]
> is there any reason to not use core.sync.* instead?

No, just go for it 😃.

**Comment by summer-alice** [comment]
Okay so we're almost there.

Both `ubuntu-20.04` LDC fail with:
```
org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/internal/Converter.d(19,15): Error: Function type does not match previously declared function with the same mangled name: `g_utf16_to_utf8`
org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/internal/Converter.d(19,15):        Previous IR type: i8* (i16*, i64, i64*, i64*, %org.eclipse.swt.internal.c.glib_object._GError**)
org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/internal/Converter.d(19,15):        New IR type:      i8* (i16*, i32, i32*, i32*, %org.eclipse.swt.internal.Converter.GError**)
/opt/hostedtoolcache/dc/ldc2-1.24.0/x64/ldc2-1.24.0-linux-x86_64/bin/ldc2 failed with exit code 1.
```

I can confirm this happens locally on my machine (Devuan) as well, didn't notice it since I don't really use LDC.

**Comment by summer-alice** [comment]
I haven't enabled testing on windows yet since I wanted to see if you had an alternative to the "duplication" used for Gtk testing.
Since I assume windows would be similar.

**Comment by jacob-carlborg** [comment]
> Both ubuntu-20.04 LDC fail with:
>
```
org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/internal/Converter.d(19,15): Error: Function type does not match previously > declared function with the same mangled name: `g_utf16_to_utf8`
org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/internal/Converter.d(19,15):        Previous IR type: i8* (i16*, i64, i64*, i64*, %org.eclipse.swt.internal.c.glib_object._GError**)
org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/internal/Converter.d(19,15):        New IR type:      i8* (i16*, i32, i32*, i32*, %org.eclipse.swt.internal.Converter.GError**)
/opt/hostedtoolcache/dc/ldc2-1.24.0/x64/ldc2-1.24.0-linux-x86_64/bin/ldc2 failed with exit code 1.
```
> I can confirm this happens locally on my machine (Devuan) as well, didn't notice it since I don't really use LDC.

So what happens here is that LDC has a typed IR, which DMD does not. That means LDC will perform some extra checks that DMD can't do. The above error happens when a function is declared multiple times with different signatures. This is a bug in the code. There are two main reason this can happen, both usually involve `extern (C)` functions.

An `extern (C)` function is **declared** in two different modules with different signatures, i.e.:

```d
module a;
extern (C) void foo(int);
```

```d
module b;
extern (C) void foo(long);
```

The above is usually caused by a bug, accidentally using the wrong signature. To fix this, change the signatures to match.

The second way this can happen, which I've seen a couple of times in DWT, is `extern (C):` is put at the beginning of a file. Then two (or more) functions with the same name are **defined** with different signatures. In this case the developer intentionally wants different signatures, i.e. overloading, but forgot the top level `extern (C):`. In that case the correct fix is to explicitly add `extern (D)` on these functions, assuming they're not used as callbacks to a C function. Example:

```d
module foo;
extern (C):
//
// a bunch of C declarations
//

// some utility functions:
void bar(int)
{
}

void bar(double)
{
}
```

**Comment by jacob-carlborg** [comment]
> Since I assume windows would be similar.

That's not necessarily true. This usually happens in the platform specific code. So it might not happen on Windows at all. Or if you fix the issue for Linux there might still be an issue for Windows.

**Comment by summer-alice** [comment]
> > Since I assume windows would be similar.
> 
> That's not necessarily true. This usually happens in the platform specific code. So it might not happen on Windows at all. Or if you fix the issue for Linux there might still be an issue for Windows.

Sorry, I forgot to mention that I was talking about the changes to `dub.sdl` in that comment (i.e. the new `unittest-gtk` configuration).

**Comment by summer-alice** [comment]
> > ```
> > org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/internal/Converter.d(19,15): Error: Function type does not match previously > declared function with the same mangled name: `g_utf16_to_utf8`
> > org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/internal/Converter.d(19,15):        Previous IR type: i8* (i16*, i64, i64*, i64*, %org.eclipse.swt.internal.c.glib_object._GError**)
> > org.eclipse.swt.gtk.linux.x86/src/org/eclipse/swt/internal/Converter.d(19,15):        New IR type:      i8* (i16*, i32, i32*, i32*, %org.eclipse.swt.internal.Converter.GError**)
> > /opt/hostedtoolcache/dc/ldc2-1.24.0/x64/ldc2-1.24.0-linux-x86_64/bin/ldc2 failed with exit code 1.
> > ```
>
> ...LDC has a typed IR, which DMD does not. That means LDC will perform some extra checks that DMD can't do.

Didn't know this.  That would explain why DMD never picked it up.

> 
> An `extern (C)` function is **declared** in two different modules with different signatures...
> [...]
> The above is usually caused by a bug, accidentally using the wrong signature. To fix this, change the signatures to match.

This did seem to be the case (see #74).

**Comment by jacob-carlborg** [comment]
What's the status of this?

**Comment by summer-alice** [comment]
> What's the status of this?

On Ubuntu the [most recent run] (are the results viewable by you?) passed for both the `base` and `dwt` packages.  Windows is currently only testing the `base` package (it passed as well). When I last tried running the `dwt` tests for Windows, the 64-bit version fails with some error number. (I didn't try 32-bit.) Also, now the Windows version doesn't have the `RGBA` class so the tests wouldn't build.

I have been spending some time setting up a Windows partition so I can try and fix these issues (better performance than the VM).

I can enable the `dwt` tests for Windows now if you'd like?

[most recent run]: https://github.com/summer-alice/dwt/actions/runs/581270386

**Comment by jacob-carlborg** [comment]
> On Ubuntu the most recent run (are the results viewable by you?) passed for both the base and dwt packages.

Awesome. And yes, the I can see the results.

> Windows is currently only testing the base package (it passed as well).

👍

> When I last tried running the dwt tests for Windows, the 64-bit version fails with some error number. (I didn't try 32-bit.)

I see.

> I have been spending some time setting up a Windows partition so I can try and fix these issues (better performance than the VM).

Perhaps you could try Wine, unless that's more complicated.

> I can enable the dwt tests for Windows now if you'd like?

No, let's not do that for now.

It's still better than before.

**Comment by summer-alice** [comment]
> > I have been spending some time setting up a Windows partition so I can try and fix these issues (better performance than the VM).
> 
> Perhaps you could try Wine, unless that's more complicated.

I don't think it would be more complicated to set up Wine, but I wanted to prevent the possibility of an issue which is either unique to Wine or not reproducible in Wine.

> > I can enable the dwt tests for Windows now if you'd like?
> 
> No, let's not do that for now.
> 
> It's still better than before.

Okay then.  I think this is pretty much complete...?  I have noticed that the `dwt:base` test doesn't specify an architecture, so I'll fix that.

If there is anything missing let me know.

**Comment by jacob-carlborg** [comment]
> I don't think it would be more complicated to set up Wine, but I wanted to prevent the possibility of an issue which is either unique to Wine or not reproducible in Wine.

That's a good point.

> Okay then. I think this is pretty much complete...? I have noticed that the dwt:base test doesn't specify an architecture, so I'll fix that.

Perfect.

> If there is anything missing let me know.

Not that I can think of right now. If there is something missing, we can always incrementally improve.

**Review by jacob-carlborg — APPROVED** [review]

**Comment by jacob-carlborg** [comment]
This all looks good. But this project is part of the Dlang CI pipeline. I need to double check that this doesn't break anything there. I need to wait a bit before merging.

**Comment by jacob-carlborg** [comment]
This is blocked by https://github.com/dlang/ci/pull/441.

**Comment by jacob-carlborg** [comment]
Thanks.

**Comment by summer-alice** [comment]
> Thanks.

Thank you.

**Pull request merged**
